src/libraries/System.Private.CoreLib/src/System/Threading/ThreadPool.cs

```diff
@@ -1058,6 +1083,221 @@ internal static void PerformWaitOrTimerCallback(_ThreadPoolWaitOrTimerCallback h
+        private static object? t_currentWorkItem;
+
+        private static ThreadPoolBlockingQueue? s_processor;
+        private static HashSet<object>? s_blockingItems;
+        private static bool s_enabled;
+
+        public static bool IsEnabled => s_enabled;
+
+        private readonly ConcurrentQueue<object> _workItems = new ConcurrentQueue<object>();
+        private int _processingThreads;
+
+        public static IThreadPoolWorkItem Enqueue(object workItem)
+        {
+            ThreadPoolBlockingQueue? processor = s_processor;
+            if (processor is null)
+            {
+                // Initialize
+                processor = new ThreadPoolBlockingQueue();
+                ThreadPoolBlockingQueue? currentProcessor = Interlocked.CompareExchange(ref s_processor, processor, null);
+                if (currentProcessor != null)
+                {
+                    processor = currentProcessor;
+                }
+            }
+
+            processor._workItems.Enqueue(workItem);
+            return processor;
+        }
+
+        void IThreadPoolWorkItem.Execute()
+        {
+            if (!IsRequired())
+            {
+                return;
+            }
+
+            Thread? currentThread = Thread.CurrentThread;
+            while (true)
+            {
+                while (_workItems.TryDequeue(out object? workItem))
+                {
+                    if (ThreadPool.EnableWorkerTracking)
+                    {
+                        ThreadPoolWorkQueue.DispatchWorkItemWithWorkerTracking(workItem, currentThread);
+                    }
+                    else if (workItem is Task task)
+                    {
+                        task.ExecuteFromThreadPool(currentThread);
+                    }
+                    else
+                    {
+                        Debug.Assert(workItem is IThreadPoolWorkItem);
+                        Unsafe.As<IThreadPoolWorkItem>(workItem).Execute();
+                    }
+                    workItem = null;
+
+                    ExecutionContext.ResetThreadPoolThread(currentThread);
+                    currentThread.ResetThreadPoolThread();
+                }
+
+                // All work done.
+                MarkCompleted();
+
+                if (_workItems.IsEmpty)
+                {
+                    // Nothing to do, exit.
+                    break;
+                }
+
+                // Is work, can we set it as active again?
+                if (!IsRequired())
+                {
+                    // Execute has been rescheduled already, exit.
+                    break;
+                }
+
+                // Is work, wasn't already scheduled so continue loop.
+            }
+        }
+
+        private bool IsRequired()
+        {
+            ThreadPool.GetMinThreads(out int min, out _);
+            // Ideally would grow with ThreadPool.ThreadCount
+            int max = Math.Max(min, Environment.ProcessorCount);
```

Comment:
The throughput isn't great for the blocking items (as they are blocked); so ideally the `max` concurrency here would grow with `ThreadPool.ThreadCount`. However; when I made this value more dynamic using that value it stopped working 😥

Reply:
e.g. even this seemed to go wrong (allowing a proc count set aside)
```csharp
ThreadPool.GetMinThreads(out int min, out _);
int max = Math.Max(1, Math.Max(min, ThreadPool.ThreadCount - Environment.ProcessorCount));
```

Reply:
Issue that was introducing is it was consuming threads as fast as the HillClimbing injecting them; limiting to `ThreadPool.ThreadCount / 2` resolves this